File: part2/examples/cpp/07_output_operator_overloading/person.cpp
#include <iostream>

using namespace std;

class person {
    
public:
    char* name;
    person(char* n) {
        name=n;
    }
    friend ostream& operator << (ostream& o, const person& p);
};

//binary operator overloading (in outer scope)
ostream& operator << (ostream& o, const person& p) {
        
        //this is not the same function called in the main funct, it depends from second argument
        cout << "that's a test" << endl;    
        
        return o << p.name;
}

int main() {
    
    person p("anna");
    cout << p << endl;
    
    return 0;
}

File: part2/examples/cpp/04_ext/Person.java
class Person {
	public void print() {
		System.out.println("person");
	}
	
	public static void main(String[] args) {
		Student s = new Student();
		Person p = s;
		
		p.print();
	}
	
}

class Student extends Person {
	public void print() {
		System.out.println("student");
	}
}
